pricegraph/fuzz/fuzz_targets/orderbook.rs

```diff
@@ -1,9 +1,50 @@
 #![no_main]
+use arbitrary::Arbitrary;
 use libfuzzer_sys::fuzz_target;
-use pricegraph::{Element, Orderbook};
+use pricegraph::{Element, Orderbook, TokenId, TokenPair};
+use std::iter::once;
+
+// Limit the maximum token id that is allowed to appear in the generated orders. Without this we can
+// make orderbook creation too slow if one order contains a large buy or sell token id.
+const MAX_TOKEN_ID: u16 = 16;
 
 // Fuzz creation and usage of Orderbook.
 
-fuzz_target!(|elements: Vec<Element>| {
-    let _ = Orderbook::from_elements(elements);
+#[derive(Arbitrary, Debug)]
+enum Operation {
+    IsOverlapping,
+    ReduceOverlappingOrders,
+    FillMarketOrders { pair: TokenPair, volume: f64 },
+}
+
+#[derive(Arbitrary, Debug)]
+struct Arguments {
+    elements: Vec<Element>,
+    operation: Operation,
+}
+
+fn largest_token_id(elements: &[Element]) -> Option<TokenId> {
+    elements
+        .iter()
+        .flat_map(|e| once(e.pair.buy).chain(once(e.pair.sell)))
+        .max()
+}
+
+fuzz_target!(|arguments: Arguments| {
+    if largest_token_id(&arguments.elements).unwrap_or(0) > MAX_TOKEN_ID {
+        return;
+    }
+
+    let mut orderbook = Orderbook::from_elements(arguments.elements);
+    match arguments.operation {
```

Comment:
Not a big deal, but could we instead of introducing the new operation type to test all operations in serial? I see that you want to fuzz the pair and volume, but I don't think we need the Operation enum (it adds bloat to the otherwise very small and easy to understand fuzzing code).

I could understand if multiple function executions were fuzzed in arbitrary order on the same instance (testing interdependency of e.g. first reducing and then filling a market order and vice versa).

Reply:
I prefer it this way.
Testing operations in serial introduces dependencies between them where the order can matter. I suspect it would make fuzzing slower and harder to find relevant branches for the fuzzer.
I wouldn't call this bloat or hard to understand either. Of course it is slightly more complex than doing it in serial but it has a specific purpose.

Reply:
Hmm, I'm kind of on the fence about this one - I would actually kind of prefer to have two targets here:
1. `fill_market_order`
2. `is_overlapping -> reduce_ovelapping_orders -> assert!(!is_overlapping)`

But maybe I'm suggesting this because I know too much about the actual implementation.

One thing to note is that `is_overlapping` takes `&self`, and since we have no interior mutability, in theory it shouldn't affect operations following it.

On the otherhand its kind of nice to be super thorough.

Reply:
If 2. is a post condition we might want to add it as a debug_assert to `reduce_overlapping_orders` instead. In general if you can think of more debug_asserts, they are helpful for the fuzzer.

Reply:
Created PR #926